consumer/src/main/java/com/flipkart/varadhi/consumer/GroupedMessageSrc.java

```diff
@@ -0,0 +1,160 @@
+                }
+                tracker.messages.add(newBatch);
+                return tracker;
+            });
+            totalInFlightMessages.addAndGet(newBatch.count());
+            if (isNewGroup.isTrue()) {
+                freeGroups.add(group.getKey());
+            }
+        }
+    }
+
+    private Map<String, List<MessageTracker>> groupMessagesByGroupId(PolledMessages<O> polledMessages) {
+        Map<String, List<MessageTracker>> groups = new HashMap<>();
+        for (PolledMessage<O> polledMessage : polledMessages) {
+            MessageTracker messageTracker = new PolledMessageTracker<>(consumer, polledMessage);
+            String groupId = messageTracker.getGroupId();
+            if (StringUtils.isBlank(groupId)) {
+                throw new IllegalStateException("Group id not found for message " + messageTracker.getMessage());
+            }
+            groups.computeIfAbsent(groupId, list -> new ArrayList<>()).add(messageTracker);
+        }
+        return groups;
+    }
+
+    private boolean hasMaxInFlightMessages() {
+        return totalInFlightMessages.get() >= maxInFlightMessages;
+    }
+
+    enum GroupStatus {
+        FREE,
+        IN_FLIGHT
+    }
+
+    static class GroupTracker {
+        GroupStatus status = GroupStatus.FREE;
+
+        LinkedList<MessageBatch> messages = new LinkedList<>();
+    }
+
+    @AllArgsConstructor
+    private class GroupedMessageTracker implements MessageTracker {
+        private final MessageTracker messageTracker;
+
+        @Override
+        public Message getMessage() {
+            return messageTracker.getMessage();
+        }
+
+        @Override
+        public void onConsumed(MessageConsumptionStatus status) {
+            messageTracker.onConsumed(status);
+            String groupId = getGroupId();
+            free(groupId, status);
+        }
+
+        private void free(String groupId, MessageConsumptionStatus status) {
+            MutableBoolean isRemaining = new MutableBoolean(false);
+            allGroupedMessages.compute(groupId, (gId, tracker) -> {
+                if (tracker == null || tracker.status == GroupStatus.FREE) {
+                    throw new IllegalStateException(String.format("Tried to free group %s: %s", gId, tracker));
+                }
+                // todo: update group consumption status in tracker?
+                var messages = tracker.messages;
+                if (!messages.isEmpty() && messages.getFirst().remaining() == 0) {
+                    messages.removeFirst();
+                }
+                if (!messages.isEmpty()) {
+                    tracker.status = GroupStatus.FREE;
+                    isRemaining.setTrue();
+                    return tracker;
+                } else {
+                    return null;
+                }
+            });
+            totalInFlightMessages.decrementAndGet();
+            if (isRemaining.isTrue()) {
+                freeGroups.addFirst(groupId);
```

Comment:
I based this logic off the internal impl. So, not aware why we do that. I think @gauravAshok can explain in more detail.

Reply:
`wouldn't round robin across all free groups make more sense` - so, the freeGroup queue "roughly" represents the order of "groups" that we have encountered. If "g1" group is present before "g2" then it is likely that "g1" will be present before in the queue as well.

So, to unblock the "batch" and their commit early, we are, in a way, prioritizing the groups. Keep processing the groups that are present in the beginning of the queue. Hence we pop from head and if there are more to process for that group, we again put it back in the head so that it can be prioritized.